README.md
./wolfram --rule x --lines y --window i --start e

x = the ruleset to use. -> 0 to 255
y = the number of lines to display. When omitted, the program never stops.
i = the number of cells to display on each line (line width). If even,the central cell is displayed in the next cell on the right. The default value is 80.
e = the generation number at which to start the display. The default value is 0.

ChangeLog.md
# Changelog for azerty

## Unreleased changes


src/Lib.hs
module Lib where

getNbr :: [String] -> Int
getNbr [] = 0 -- ERROR
getNbr (x:xs) = read x :: Int

myLength :: String -> Int
myLength [] = 0
myLength (_:xs) = 1 + myLength xs

myAppend :: [a] -> [a] -> [a]
myAppend (x:xs) c = x : myAppend xs c
myAppend [] x = x

myReverse :: [a] -> [a]
myReverse [] = []
myReverse (x:xs) = myAppend (myReverse xs) [x]

bintoint :: Int -> Int -> Int
bintoint a b | a >= 100 = bintoint (a - 100) (b + 4)
             | a >= 10 = bintoint (a - 10) (b + 2)
             | a >= 1 = bintoint (a - 1) (b + 1)
             | otherwise = b

remp :: String -> Int -> String -> String
remp [] _ stock = stock
remp (x:xs) b stock | b-9 == 0 = remp xs b (stock ++ [x])
                    | otherwise = remp (x:xs) (b+1) (stock ++ "0")

intobin :: Int -> Int -> String
intobin a c | a >= 128 = intobin (a - 128) (c + 10000000)
            | a >= 64 = intobin (a - 64) (c + 1000000)
            | a >= 32 = intobin (a - 32) (c + 100000)
            | a >= 16 = intobin (a - 16) (c + 10000)
            | a >= 8 = intobin (a - 8) (c + 1000)
            | a >= 4 = intobin (a - 4) (c + 100)
            | a >= 2 = intobin (a - 2) (c + 10)
            | a >= 1 = intobin (a - 1) (c + 1)
            | otherwise = remp (show c) (myLength (show c)) ""

stri :: Int -> String -> Int -> Bool
stri _ [] _ = False
stri a (x:xs) b | a == b && x == '1' = True
                | a /= b = stri a xs (b+1)
                | otherwise = False

see :: Int -> String -> Bool
see a b | stri a b 0 = True
        | otherwise = False

look :: String -> String -> Bool
look str a | see (8 - bintoint (read str :: Int) 0) a = True
           | otherwise = False

func :: String -> Int -> String -> String
func [] _ _ = ""
func (x:xs) a str | a-3 >= 0 = func xs (a-1) str
                  | a /= 0 = func xs (a-1) str ++ [x]
                  | otherwise = str ++ [x]


.stack-work/ghci/fe4ef255/cabal_macros.h
/* DO NOT EDIT: This file is automatically generated by Cabal */

/* package azerty-0.1.0.0 */
#ifndef VERSION_azerty
#define VERSION_azerty "0.1.0.0"
#endif /* VERSION_azerty */
#ifndef MIN_VERSION_azerty
#define MIN_VERSION_azerty(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  1 || \
  (major1) == 0 && (major2) == 1 && (minor) <= 0)
#endif /* MIN_VERSION_azerty */

/* package base-4.13.0.0 */
#ifndef VERSION_base
#define VERSION_base "4.13.0.0"
#endif /* VERSION_base */
#ifndef MIN_VERSION_base
#define MIN_VERSION_base(major1,major2,minor) (\
  (major1) <  4 || \
  (major1) == 4 && (major2) <  13 || \
  (major1) == 4 && (major2) == 13 && (minor) <= 0)
#endif /* MIN_VERSION_base */

/* tool alex-3.2.4 */
#ifndef TOOL_VERSION_alex
#define TOOL_VERSION_alex "3.2.4"
#endif /* TOOL_VERSION_alex */
#ifndef MIN_TOOL_VERSION_alex
#define MIN_TOOL_VERSION_alex(major1,major2,minor) (\
  (major1) <  3 || \
  (major1) == 3 && (major2) <  2 || \
  (major1) == 3 && (major2) == 2 && (minor) <= 4)
#endif /* MIN_TOOL_VERSION_alex */

/* tool gcc-9 */
#ifndef TOOL_VERSION_gcc
#define TOOL_VERSION_gcc "9"
#endif /* TOOL_VERSION_gcc */
#ifndef MIN_TOOL_VERSION_gcc
#define MIN_TOOL_VERSION_gcc(major1,major2,minor) (\
  (major1) <  9 || \
  (major1) == 9 && (major2) <  0 || \
  (major1) == 9 && (major2) == 0 && (minor) <= 0)
#endif /* MIN_TOOL_VERSION_gcc */

/* tool ghc-8.8.4 */
#ifndef TOOL_VERSION_ghc
#define TOOL_VERSION_ghc "8.8.4"
#endif /* TOOL_VERSION_ghc */
#ifndef MIN_TOOL_VERSION_ghc
#define MIN_TOOL_VERSION_ghc(major1,major2,minor) (\
  (major1) <  8 || \
  (major1) == 8 && (major2) <  8 || \
  (major1) == 8 && (major2) == 8 && (minor) <= 4)
#endif /* MIN_TOOL_VERSION_ghc */

/* tool ghc-pkg-8.8.4 */
#ifndef TOOL_VERSION_ghc_pkg
#define TOOL_VERSION_ghc_pkg "8.8.4"
#endif /* TOOL_VERSION_ghc_pkg */
#ifndef MIN_TOOL_VERSION_ghc_pkg
#define MIN_TOOL_VERSION_ghc_pkg(major1,major2,minor) (\
  (major1) <  8 || \
  (major1) == 8 && (major2) <  8 || \
  (major1) == 8 && (major2) == 8 && (minor) <= 4)
#endif /* MIN_TOOL_VERSION_ghc_pkg */

/* tool haddock-2.23.0 */
#ifndef TOOL_VERSION_haddock
#define TOOL_VERSION_haddock "2.23.0"
#endif /* TOOL_VERSION_haddock */
#ifndef MIN_TOOL_VERSION_haddock
#define MIN_TOOL_VERSION_haddock(major1,major2,minor) (\
  (major1) <  2 || \
  (major1) == 2 && (major2) <  23 || \
  (major1) == 2 && (major2) == 23 && (minor) <= 0)
#endif /* MIN_TOOL_VERSION_haddock */

/* tool happy-1.19.11 */
#ifndef TOOL_VERSION_happy
#define TOOL_VERSION_happy "1.19.11"
#endif /* TOOL_VERSION_happy */
#ifndef MIN_TOOL_VERSION_happy
#define MIN_TOOL_VERSION_happy(major1,major2,minor) (\
  (major1) <  1 || \
  (major1) == 1 && (major2) <  19 || \
  (major1) == 1 && (major2) == 19 && (minor) <= 11)
#endif /* MIN_TOOL_VERSION_happy */

/* tool hpc-0.67 */
#ifndef TOOL_VERSION_hpc
#define TOOL_VERSION_hpc "0.67"
#endif /* TOOL_VERSION_hpc */
#ifndef MIN_TOOL_VERSION_hpc
#define MIN_TOOL_VERSION_hpc(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  67 || \
  (major1) == 0 && (major2) == 67 && (minor) <= 0)
#endif /* MIN_TOOL_VERSION_hpc */

/* tool hsc2hs-0.68.7 */
#ifndef TOOL_VERSION_hsc2hs
#define TOOL_VERSION_hsc2hs "0.68.7"
#endif /* TOOL_VERSION_hsc2hs */
#ifndef MIN_TOOL_VERSION_hsc2hs
#define MIN_TOOL_VERSION_hsc2hs(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  68 || \
  (major1) == 0 && (major2) == 68 && (minor) <= 7)
#endif /* MIN_TOOL_VERSION_hsc2hs */

/* tool hscolour-1.24 */
#ifndef TOOL_VERSION_hscolour
#define TOOL_VERSION_hscolour "1.24"
#endif /* TOOL_VERSION_hscolour */
#ifndef MIN_TOOL_VERSION_hscolour
#define MIN_TOOL_VERSION_hscolour(major1,major2,minor) (\
  (major1) <  1 || \
  (major1) == 1 && (major2) <  24 || \
  (major1) == 1 && (major2) == 24 && (minor) <= 0)
#endif /* MIN_TOOL_VERSION_hscolour */

/* tool pkg-config-1.6.3 */
#ifndef TOOL_VERSION_pkg_config
#define TOOL_VERSION_pkg_config "1.6.3"
#endif /* TOOL_VERSION_pkg_config */
#ifndef MIN_TOOL_VERSION_pkg_config
#define MIN_TOOL_VERSION_pkg_config(major1,major2,minor) (\
  (major1) <  1 || \
  (major1) == 1 && (major2) <  6 || \
  (major1) == 1 && (major2) == 6 && (minor) <= 3)
#endif /* MIN_TOOL_VERSION_pkg_config */

/* tool runghc-8.8.4 */
#ifndef TOOL_VERSION_runghc
#define TOOL_VERSION_runghc "8.8.4"
#endif /* TOOL_VERSION_runghc */
#ifndef MIN_TOOL_VERSION_runghc
#define MIN_TOOL_VERSION_runghc(major1,major2,minor) (\
  (major1) <  8 || \
  (major1) == 8 && (major2) <  8 || \
  (major1) == 8 && (major2) == 8 && (minor) <= 4)
#endif /* MIN_TOOL_VERSION_runghc */

/* tool strip-2.32 */
#ifndef TOOL_VERSION_strip
#define TOOL_VERSION_strip "2.32"
#endif /* TOOL_VERSION_strip */
#ifndef MIN_TOOL_VERSION_strip
#define MIN_TOOL_VERSION_strip(major1,major2,minor) (\
  (major1) <  2 || \
  (major1) == 2 && (major2) <  32 || \
  (major1) == 2 && (major2) == 32 && (minor) <= 0)
#endif /* MIN_TOOL_VERSION_strip */

#ifndef CURRENT_PACKAGE_KEY
#define CURRENT_PACKAGE_KEY "azerty-0.1.0.0-FuAFKh8LtNeBRwFFvHVDim"
#endif /* CURRENT_PACKAGE_KEY */
#ifndef CURRENT_COMPONENT_ID
#define CURRENT_COMPONENT_ID "azerty-0.1.0.0-FuAFKh8LtNeBRwFFvHVDim"
#endif /* CURRENT_COMPONENT_ID */
#ifndef CURRENT_PACKAGE_VERSION
#define CURRENT_PACKAGE_VERSION "0.1.0.0"
#endif /* CURRENT_PACKAGE_VERSION */

#undef CURRENT_PACKAGE_KEY
#undef CURRENT_COMPONENT_ID
/* DO NOT EDIT: This file is automatically generated by Cabal */

/* package azerty-0.1.0.0 */
#ifndef VERSION_azerty
#define VERSION_azerty "0.1.0.0"
#endif /* VERSION_azerty */
#ifndef MIN_VERSION_azerty
#define MIN_VERSION_azerty(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  1 || \
  (major1) == 0 && (major2) == 1 && (minor) <= 0)
#endif /* MIN_VERSION_azerty */

/* package azerty-0.1.0.0 */
#ifndef VERSION_azerty
#define VERSION_azerty "0.1.0.0"
#endif /* VERSION_azerty */
#ifndef MIN_VERSION_azerty
#define MIN_VERSION_azerty(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  1 || \
  (major1) == 0 && (major2) == 1 && (minor) <= 0)
#endif /* MIN_VERSION_azerty */

/* package base-4.13.0.0 */
#ifndef VERSION_base
#define VERSION_base "4.13.0.0"
#endif /* VERSION_base */
#ifndef MIN_VERSION_base
#define MIN_VERSION_base(major1,major2,minor) (\
  (major1) <  4 || \
  (major1) == 4 && (major2) <  13 || \
  (major1) == 4 && (major2) == 13 && (minor) <= 0)
#endif /* MIN_VERSION_base */

/* tool alex-3.2.4 */
#ifndef TOOL_VERSION_alex
#define TOOL_VERSION_alex "3.2.4"
#endif /* TOOL_VERSION_alex */
#ifndef MIN_TOOL_VERSION_alex
#define MIN_TOOL_VERSION_alex(major1,major2,minor) (\
  (major1) <  3 || \
  (major1) == 3 && (major2) <  2 || \
  (major1) == 3 && (major2) == 2 && (minor) <= 4)
#endif /* MIN_TOOL_VERSION_alex */

/* tool gcc-9 */
#ifndef TOOL_VERSION_gcc
#define TOOL_VERSION_gcc "9"
#endif /* TOOL_VERSION_gcc */
#ifndef MIN_TOOL_VERSION_gcc
#define MIN_TOOL_VERSION_gcc(major1,major2,minor) (\
  (major1) <  9 || \
  (major1) == 9 && (major2) <  0 || \
  (major1) == 9 && (major2) == 0 && (minor) <= 0)
#endif /* MIN_TOOL_VERSION_gcc */

/* tool ghc-8.8.4 */
#ifndef TOOL_VERSION_ghc
#define TOOL_VERSION_ghc "8.8.4"
#endif /* TOOL_VERSION_ghc */
#ifndef MIN_TOOL_VERSION_ghc
#define MIN_TOOL_VERSION_ghc(major1,major2,minor) (\
  (major1) <  8 || \
  (major1) == 8 && (major2) <  8 || \
  (major1) == 8 && (major2) == 8 && (minor) <= 4)
#endif /* MIN_TOOL_VERSION_ghc */

/* tool ghc-pkg-8.8.4 */
#ifndef TOOL_VERSION_ghc_pkg
#define TOOL_VERSION_ghc_pkg "8.8.4"
#endif /* TOOL_VERSION_ghc_pkg */
#ifndef MIN_TOOL_VERSION_ghc_pkg
#define MIN_TOOL_VERSION_ghc_pkg(major1,major2,minor) (\
  (major1) <  8 || \
  (major1) == 8 && (major2) <  8 || \
  (major1) == 8 && (major2) == 8 && (minor) <= 4)
#endif /* MIN_TOOL_VERSION_ghc_pkg */

/* tool haddock-2.23.0 */
#ifndef TOOL_VERSION_haddock
#define TOOL_VERSION_haddock "2.23.0"
#endif /* TOOL_VERSION_haddock */
#ifndef MIN_TOOL_VERSION_haddock
#define MIN_TOOL_VERSION_haddock(major1,major2,minor) (\
  (major1) <  2 || \
  (major1) == 2 && (major2) <  23 || \
  (major1) == 2 && (major2) == 23 && (minor) <= 0)
#endif /* MIN_TOOL_VERSION_haddock */

/* tool happy-1.19.11 */
#ifndef TOOL_VERSION_happy
#define TOOL_VERSION_happy "1.19.11"
#endif /* TOOL_VERSION_happy */
#ifndef MIN_TOOL_VERSION_happy
#define MIN_TOOL_VERSION_happy(major1,major2,minor) (\
  (major1) <  1 || \
  (major1) == 1 && (major2) <  19 || \
  (major1) == 1 && (major2) == 19 && (minor) <= 11)
#endif /* MIN_TOOL_VERSION_happy */

/* tool hpc-0.67 */
#ifndef TOOL_VERSION_hpc
#define TOOL_VERSION_hpc "0.67"
#endif /* TOOL_VERSION_hpc */
#ifndef MIN_TOOL_VERSION_hpc
#define MIN_TOOL_VERSION_hpc(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  67 || \
  (major1) == 0 && (major2) == 67 && (minor) <= 0)
#endif /* MIN_TOOL_VERSION_hpc */

/* tool hsc2hs-0.68.7 */
#ifndef TOOL_VERSION_hsc2hs
#define TOOL_VERSION_hsc2hs "0.68.7"
#endif /* TOOL_VERSION_hsc2hs */
#ifndef MIN_TOOL_VERSION_hsc2hs
#define MIN_TOOL_VERSION_hsc2hs(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  68 || \
  (major1) == 0 && (major2) == 68 && (minor) <= 7)
#endif /* MIN_TOOL_VERSION_hsc2hs */

/* tool hscolour-1.24 */
#ifndef TOOL_VERSION_hscolour
#define TOOL_VERSION_hscolour "1.24"
#endif /* TOOL_VERSION_hscolour */
#ifndef MIN_TOOL_VERSION_hscolour
#define MIN_TOOL_VERSION_hscolour(major1,major2,minor) (\
  (major1) <  1 || \
  (major1) == 1 && (major2) <  24 || \
  (major1) == 1 && (major2) == 24 && (minor) <= 0)
#endif /* MIN_TOOL_VERSION_hscolour */

/* tool pkg-config-1.6.3 */
#ifndef TOOL_VERSION_pkg_config
#define TOOL_VERSION_pkg_config "1.6.3"
#endif /* TOOL_VERSION_pkg_config */
#ifndef MIN_TOOL_VERSION_pkg_config
#define MIN_TOOL_VERSION_pkg_config(major1,major2,minor) (\
  (major1) <  1 || \
  (major1) == 1 && (major2) <  6 || \
  (major1) == 1 && (major2) == 6 && (minor) <= 3)
#endif /* MIN_TOOL_VERSION_pkg_config */

/* tool runghc-8.8.4 */
#ifndef TOOL_VERSION_runghc
#define TOOL_VERSION_runghc "8.8.4"
#endif /* TOOL_VERSION_runghc */
#ifndef MIN_TOOL_VERSION_runghc
#define MIN_TOOL_VERSION_runghc(major1,major2,minor) (\
  (major1) <  8 || \
  (major1) == 8 && (major2) <  8 || \
  (major1) == 8 && (major2) == 8 && (minor) <= 4)
#endif /* MIN_TOOL_VERSION_runghc */

/* tool strip-2.32 */
#ifndef TOOL_VERSION_strip
#define TOOL_VERSION_strip "2.32"
#endif /* TOOL_VERSION_strip */
#ifndef MIN_TOOL_VERSION_strip
#define MIN_TOOL_VERSION_strip(major1,major2,minor) (\
  (major1) <  2 || \
  (major1) == 2 && (major2) <  32 || \
  (major1) == 2 && (major2) == 32 && (minor) <= 0)
#endif /* MIN_TOOL_VERSION_strip */

#ifndef CURRENT_COMPONENT_ID
#define CURRENT_COMPONENT_ID "azerty-0.1.0.0-2aeEObBGPknJ47eE8Wu36t-azerty-exe"
#endif /* CURRENT_COMPONENT_ID */
#ifndef CURRENT_PACKAGE_VERSION
#define CURRENT_PACKAGE_VERSION "0.1.0.0"
#endif /* CURRENT_PACKAGE_VERSION */

#undef CURRENT_PACKAGE_KEY
#undef CURRENT_COMPONENT_ID
/* DO NOT EDIT: This file is automatically generated by Cabal */

/* package azerty-0.1.0.0 */
#ifndef VERSION_azerty
#define VERSION_azerty "0.1.0.0"
#endif /* VERSION_azerty */
#ifndef MIN_VERSION_azerty
#define MIN_VERSION_azerty(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  1 || \
  (major1) == 0 && (major2) == 1 && (minor) <= 0)
#endif /* MIN_VERSION_azerty */

/* package azerty-0.1.0.0 */
#ifndef VERSION_azerty
#define VERSION_azerty "0.1.0.0"
#endif /* VERSION_azerty */
#ifndef MIN_VERSION_azerty
#define MIN_VERSION_azerty(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  1 || \
  (major1) == 0 && (major2) == 1 && (minor) <= 0)
#endif /* MIN_VERSION_azerty */

/* package base-4.13.0.0 */
#ifndef VERSION_base
#define VERSION_base "4.13.0.0"
#endif /* VERSION_base */
#ifndef MIN_VERSION_base
#define MIN_VERSION_base(major1,major2,minor) (\
  (major1) <  4 || \
  (major1) == 4 && (major2) <  13 || \
  (major1) == 4 && (major2) == 13 && (minor) <= 0)
#endif /* MIN_VERSION_base */

/* tool alex-3.2.4 */
#ifndef TOOL_VERSION_alex
#define TOOL_VERSION_alex "3.2.4"
#endif /* TOOL_VERSION_alex */
#ifndef MIN_TOOL_VERSION_alex
#define MIN_TOOL_VERSION_alex(major1,major2,minor) (\
  (major1) <  3 || \
  (major1) == 3 && (major2) <  2 || \
  (major1) == 3 && (major2) == 2 && (minor) <= 4)
#endif /* MIN_TOOL_VERSION_alex */

/* tool gcc-9 */
#ifndef TOOL_VERSION_gcc
#define TOOL_VERSION_gcc "9"
#endif /* TOOL_VERSION_gcc */
#ifndef MIN_TOOL_VERSION_gcc
#define MIN_TOOL_VERSION_gcc(major1,major2,minor) (\
  (major1) <  9 || \
  (major1) == 9 && (major2) <  0 || \
  (major1) == 9 && (major2) == 0 && (minor) <= 0)
#endif /* MIN_TOOL_VERSION_gcc */

/* tool ghc-8.8.4 */
#ifndef TOOL_VERSION_ghc
#define TOOL_VERSION_ghc "8.8.4"
#endif /* TOOL_VERSION_ghc */
#ifndef MIN_TOOL_VERSION_ghc
#define MIN_TOOL_VERSION_ghc(major1,major2,minor) (\
  (major1) <  8 || \
  (major1) == 8 && (major2) <  8 || \
  (major1) == 8 && (major2) == 8 && (minor) <= 4)
#endif /* MIN_TOOL_VERSION_ghc */

/* tool ghc-pkg-8.8.4 */
#ifndef TOOL_VERSION_ghc_pkg
#define TOOL_VERSION_ghc_pkg "8.8.4"
#endif /* TOOL_VERSION_ghc_pkg */
#ifndef MIN_TOOL_VERSION_ghc_pkg
#define MIN_TOOL_VERSION_ghc_pkg(major1,major2,minor) (\
  (major1) <  8 || \
  (major1) == 8 && (major2) <  8 || \
  (major1) == 8 && (major2) == 8 && (minor) <= 4)
#endif /* MIN_TOOL_VERSION_ghc_pkg */

/* tool haddock-2.23.0 */
#ifndef TOOL_VERSION_haddock
#define TOOL_VERSION_haddock "2.23.0"
#endif /* TOOL_VERSION_haddock */
#ifndef MIN_TOOL_VERSION_haddock
#define MIN_TOOL_VERSION_haddock(major1,major2,minor) (\
  (major1) <  2 || \
  (major1) == 2 && (major2) <  23 || \
  (major1) == 2 && (major2) == 23 && (minor) <= 0)
#endif /* MIN_TOOL_VERSION_haddock */

/* tool happy-1.19.11 */
#ifndef TOOL_VERSION_happy
#define TOOL_VERSION_happy "1.19.11"
#endif /* TOOL_VERSION_happy */
#ifndef MIN_TOOL_VERSION_happy
#define MIN_TOOL_VERSION_happy(major1,major2,minor) (\
  (major1) <  1 || \
  (major1) == 1 && (major2) <  19 || \
  (major1) == 1 && (major2) == 19 && (minor) <= 11)
#endif /* MIN_TOOL_VERSION_happy */

/* tool hpc-0.67 */
#ifndef TOOL_VERSION_hpc
#define TOOL_VERSION_hpc "0.67"
#endif /* TOOL_VERSION_hpc */
#ifndef MIN_TOOL_VERSION_hpc
#define MIN_TOOL_VERSION_hpc(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  67 || \
  (major1) == 0 && (major2) == 67 && (minor) <= 0)
#endif /* MIN_TOOL_VERSION_hpc */

/* tool hsc2hs-0.68.7 */
#ifndef TOOL_VERSION_hsc2hs
#define TOOL_VERSION_hsc2hs "0.68.7"
#endif /* TOOL_VERSION_hsc2hs */
#ifndef MIN_TOOL_VERSION_hsc2hs
#define MIN_TOOL_VERSION_hsc2hs(major1,major2,minor) (\
  (major1) <  0 || \
  (major1) == 0 && (major2) <  68 || \
  (major1) == 0 && (major2) == 68 && (minor) <= 7)
#endif /* MIN_TOOL_VERSION_hsc2hs */

/* tool hscolour-1.24 */
#ifndef TOOL_VERSION_hscolour
#define TOOL_VERSION_hscolour "1.24"
#endif /* TOOL_VERSION_hscolour */
#ifndef MIN_TOOL_VERSION_hscolour
#define MIN_TOOL_VERSION_hscolour(major1,major2,minor) (\
  (major1) <  1 || \
  (major1) == 1 && (major2) <  24 || \
  (major1) == 1 && (major2) == 24 && (minor) <= 0)
#endif /* MIN_TOOL_VERSION_hscolour */

/* tool pkg-config-1.6.3 */
#ifndef TOOL_VERSION_pkg_config
#define TOOL_VERSION_pkg_config "1.6.3"
#endif /* TOOL_VERSION_pkg_config */
#ifndef MIN_TOOL_VERSION_pkg_config
#define MIN_TOOL_VERSION_pkg_config(major1,major2,minor) (\
  (major1) <  1 || \
  (major1) == 1 && (major2) <  6 || \
  (major1) == 1 && (major2) == 6 && (minor) <= 3)
#endif /* MIN_TOOL_VERSION_pkg_config */

/* tool runghc-8.8.4 */
#ifndef TOOL_VERSION_runghc
#define TOOL_VERSION_runghc "8.8.4"
#endif /* TOOL_VERSION_runghc */
#ifndef MIN_TOOL_VERSION_runghc
#define MIN_TOOL_VERSION_runghc(major1,major2,minor) (\
  (major1) <  8 || \
  (major1) == 8 && (major2) <  8 || \
  (major1) == 8 && (major2) == 8 && (minor) <= 4)
#endif /* MIN_TOOL_VERSION_runghc */

/* tool strip-2.32 */
#ifndef TOOL_VERSION_strip
#define TOOL_VERSION_strip "2.32"
#endif /* TOOL_VERSION_strip */
#ifndef MIN_TOOL_VERSION_strip
#define MIN_TOOL_VERSION_strip(major1,major2,minor) (\
  (major1) <  2 || \
  (major1) == 2 && (major2) <  32 || \
  (major1) == 2 && (major2) == 32 && (minor) <= 0)
#endif /* MIN_TOOL_VERSION_strip */

#ifndef CURRENT_COMPONENT_ID
#define CURRENT_COMPONENT_ID "azerty-0.1.0.0-23woXsMDW1PIf19EJ8eTsF-azerty-test"
#endif /* CURRENT_COMPONENT_ID */
#ifndef CURRENT_PACKAGE_VERSION
#define CURRENT_PACKAGE_VERSION "0.1.0.0"
#endif /* CURRENT_PACKAGE_VERSION */

#undef CURRENT_PACKAGE_KEY
#undef CURRENT_COMPONENT_ID


app/Main.hs
module Main where

import System.Exit
import System.Environment
import Check
import Lib

data Config = Cfg {
rule :: Int,
start :: Int,
line :: Int,
window :: Int,
move :: Int,
precedline :: String,
binary :: String
} deriving Show

endLine :: Config -> Int -> IO()
endLine cfg c | window cfg+1 > 1 = putStr " " >> endLine (cfg {window = window cfg - 1}) (c+1)
              | otherwise = putStr "\n"

ver :: Int -> String -> Int
ver 0 _ = 2
ver 1 _ = 2
ver _ str | myLength str >= 2 = 3
          | otherwise = 2

verif :: Int -> Int -> Config -> Bool
verif d c cfg | c+1 > (myLength (precedline cfg) - d) `div` 2 && c+1 < myLength (precedline cfg) - (myLength (precedline cfg) - d) `div` 2 = True
              | otherwise = False

printLine :: String -> Config -> Int -> Bool -> Int -> Int -> IO()
printLine (x:xs) cfg a b 0 d = printLine xs cfg a b 1 d
printLine (x:xs) cfg a b 1 d = printLine xs cfg a b 2 d
printLine [] cfg _ True c d =  endLine cfg c
printLine b cfg a False _ d | start cfg > 0 = putStr ""
                            | window cfg > a = putStr " " >> printLine (binary cfg) (cfg {window = window cfg - 1}) a False 0 d
                            | otherwise = printLine (precedline cfg) (cfg {window = window cfg - 1}) a True 0 d
printLine (x:xs) cfg a b c d | x == '0' && ver c xs /= 2 && verif d c cfg = putStr " " >> printLine xs cfg a True (c+1) d
                             | x == '1' && ver c xs /= 2 && verif d c cfg = putStr "*" >> printLine xs cfg a True (c+1) d
                             | otherwise = printLine xs cfg a b (c+1) d

check :: Config -> Int -> String
check cfg a | a < 2 || a >= myLength (precedline cfg) = "0"
            | look (myReverse(func (precedline cfg) a "")) (binary cfg) = "1"
            | otherwise = "0"

changeLine :: Config -> String -> Int -> String
changeLine cfg str a | a == myLength (precedline cfg) + 2 = str
                     | otherwise = changeLine cfg (str ++ check cfg a) (a + 1)

algo :: Config -> Bool -> Int -> Int -> IO()
algo cfg a b c | not a = printLine (precedline cfg) cfg (window cfg `div` 2) False 0 c >> algo cfg {window = window cfg -2} True b c
           | b /= 1 = algo (cfg {start = start cfg -1, line = line cfg - 1, precedline = changeLine cfg "" 0}) False (b-1) c
           | otherwise = exitSuccess

checkAll :: Config -> Bool
checkAll cfg | rule cfg < 0 || rule cfg > 255 = False
             | start cfg < 0 = False
             | window cfg < 0 = False
             | line cfg == -111444 || line cfg > 0 = True
             | otherwise = False

checkLigne :: [String] -> Config -> Bool -> Config
checkLigne [] cfg False = cfg {rule = -1}
checkLigne [] cfg True | not (checkAll cfg) = cfg {rule = -1}
                       | otherwise = cfg
checkLigne (x:xs) cfg b | not b && not (checkNbr x False) = cfg {rule = -1}
           | b && checkNbr x False = cfg {rule = -1}
           | checkNbr x False && not b = checkLigne xs cfg True
           | x == "--rule" = checkLigne xs cfg{rule = getNbr xs} False
           | x == "--start" = checkLigne xs cfg{start = getNbr xs} False
           | x == "--lines" = checkLigne xs cfg{line = getNbr xs} False
           | x == "--window" = checkLigne xs cfg{window = getNbr xs} False
           | otherwise = cfg {rule = -1}

checkArgs :: [String] -> Config -> Bool
checkArgs [] _ = True
checkArgs [""] _ = False
checkArgs ("":_) _ = False
checkArgs arg cfg | rule cfg == -1 = False
                  | otherwise = True

isArgs :: [String] -> Bool
isArgs [] = False
isArgs [[]] = False
isArgs _ = True

errorHandleur :: [String] -> Bool
errorHandleur arg | not (isArgs arg) = False
                  | otherwise = True

wolfram :: [String] -> Config -> IO()
wolfram arg cfg | not (checkArgs arg cfg) = putStrLn "KO" >> exitWith(ExitFailure 84)
                 | otherwise = algo cfg {binary = intobin (rule cfg) 0} False (start cfg + line cfg) (window cfg)

main :: IO()
main = do
    let cfg = Cfg {rule = -1, start = 0, line = -111444, window = 80, move = 0, precedline = "00100", binary = ""}
    arg <- getArgs
    if not (errorHandleur arg)
       then putStrLn "No ruleset -- KO" >> exitWith(ExitFailure 84)
    else wolfram arg (checkLigne arg cfg True)
